### PR TITLE
Fix talk tags page when navigating between projects

### PR DESCRIPTION
**Pull request by srallen**
Staging branch URL: https://pr-5690.pfe-preview.zooniverse.org

Fixes issue described on Talk: https://www.zooniverse.org/talk/17/1353127

The Talk tags page isn't loading the correct tags for the project context it is in when navigating from one project to another. The issue is that the original page code didn't account for volunteers linking to another project's tag page, so it doesn't validate that the params in the URL match the project being pass via props and doesn't request again for tags if the project in props changes. So, when navigating from the Chimp & See Talk thread to Camera CATalogue's tag page, what happens is that C&S tags are requested for:

Screen Shot 2020-04-24 at 3 44 20 PM

Screen Shot 2020-04-24 at 3 44 44 PM

(project id 7850 is C&S)

Here is what the page should be for Camera CATalogue:

Screen Shot 2020-04-24 at 3 18 29 PM

My PR fixes this by refactoring to javascript, then adding in a couple of checks. One in `componentDidMount` to make sure the URL params actually match the slug on the project passed in props before making the request for tags. Then, adding a condition to `componentWillReceiveProps` to request for tags if the project has changed.

The best way to manually test this I've found is to go to the Talk comment in C&S on the branched deployment:

https://fix-talk-tags-page.pfe-preview.zooniverse.org/projects/sassydumbledore/chimp-and-see/talk/2181/1300357?comment=2198306&env=production

Edit the `href` of the link to Camera CATalogue to be `https://fix-talk-tags-page.pfe-preview.zooniverse.org/projects/panthera-research/camera-catalogue/talk/tags/nightjar?env=production`. Then click on it and navigate to the tags page for Camera CATalogue and confirm you see the results for the correct project. The network tab should show a request to the Talk API for `/popular` with a param `section=project-825` (this is Camera CATalogue). 

# Required Manual Testing

- [x] Does the non-logged in home page render correctly?
- [x] Does the logged in home page render correctly?
- [x] Does the projects page render correctly?
- [x] Can you load project home pages?
- [x] Can you load the classification page?
- [x] Can you submit a classification?
- [x] Does talk load correctly?
- [x] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

**Comment by srallen** [comment]
Well, adding a smoke test bit me, so I'll fix the broken test early next week.

**Comment by eatyourgreens** [comment]
The changes sound good, but `componentWillReceiveProps` is deprecated, and flagged as unsafe. https://reactjs.org/docs/react-component.html

Can the component use `componentDidUpdate`?

**Comment by srallen** [comment]
@eatyourgreens if you're going to leave review-like comments, could you submit them as an actual review so I know if any of the requested changes are blocking? I certainly could do more refactoring whether it is updating the lifecycle methods or converting to a functional component with hooks, but I want to keep this scoped to the bug fix. The bug fix was easier to reason about converting to javascript since I haven't worked with coffeescript in a while. Any other changes should necessitate adding real tests to make sure regressions aren't introduced.

File: app/talk/tags.jsx [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    const differentProject = nextProps.project !== this.props.project
+
+    if (pageChanged || differentTag || differentProject) {
+      const nextPage = (differentTag) ? 1 : nextProps.location.query.page
+      this.getTags(nextPage, nextProps.params.tag, nextProps.project)
+    }
+  }
+
+  getTags(page = this.props.location.query.page, name = this.props.params.tag, project = this.props.project) {
+    page || (page = 1);
```
Comment (wgranger) [inline app/talk/tags.jsx:48]: Is this assigning page to 1 in the case it is undefined from `this.props.location.query.page`?

File: app/talk/tags.jsx [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  componentWillReceiveProps(nextProps) {
+    const pageChanged = nextProps.location.query.page !== this.props.location.query.page
+    const differentTag = nextProps.params.tag !== this.props.params.tag
+    const differentProject = nextProps.project !== this.props.project
+
+    if (pageChanged || differentTag || differentProject) {
+      const nextPage = (differentTag) ? 1 : nextProps.location.query.page
+      this.getTags(nextPage, nextProps.params.tag, nextProps.project)
+    }
+  }
```
Comment (eatyourgreens) [inline app/talk/tags.jsx:45]:
```suggestion
  componentDidUpdate(prevProps) {
    const pageChanged = prevProps.location.query.page !== this.props.location.query.page
    const differentTag = prevProps.params.tag !== this.props.params.tag
    const differentProject = prevProps.project !== this.props.project

    if (pageChanged || differentTag || differentProject) {
      const nextPage = (differentTag) ? 1 : this.props.location.query.page
      this.getTags(nextPage, this.props.params.tag, this.props.project)
    }
  }
```

**Comment by eatyourgreens** [comment]
I've added a suggested change to implement `componentDidUpdate`. We've dropped `componentWillReceiveProps` in other coffeescript conversions, and it seems a small change to implement here.

File: app/talk/tags.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (srallen) [reply app/talk/tags.jsx:48]: Yes, and I borrowed the suggested syntax from [coffeescript.org](https://coffeescript.org/#try:%20%20getTags%3A%20(page%20%3D%20%40props.location.query.page%2C%20name%20%3D%20%40props.params.tag)%20-%3E%0A%20%20%20%20page%20or%3D%201%0A%20%20%20%20taggable_type%20%3D%20'Subject'%0A%20%20%20%20section%20%3D%20%22project-%23%7B%20%40props.project.id%20%7D%22%0A%0A%20%20%20%20talkClient.type('tags%2Fpopular').get(%7Bpage%2C%20taggable_type%2C%20section%2C%20name%7D).then%20(tags)%20%3D%3E%0A%20%20%20%20%20%20meta%20%3D%20tags%5B0%5D%3F.getMeta()%0A%20%20%20%20%20%20Promise.all%20tags.map%20(tag)%20%3D%3E%0A%20%20%20%20%20%20%20%20apiClient.type('subjects').get(tag.taggable_id.toString()).then%20(subject)%20%3D%3E%0A%20%20%20%20%20%20%20%20%20%20taggable_id%20%3D%20subject.id%0A%20%20%20%20%20%20%20%20%20%20talkClient.type('tags%2Fpopular').get(%7Btaggable_type%2C%20taggable_id%7D).then%20(subjectTags)%20%3D%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20tag.update%20%7Bsubject%2C%20subjectTags%7D%0A%20%20%20%20%20%20.then%20(tags)%20%3D%3E%0A%20%20%20%20%20%20%20%20%40setState%20%7Btags%2C%20meta%7D)

File: app/talk/tags.jsx (hunk 2 above)
Comment (srallen) [reply app/talk/tags.jsx:45]: I'm happy to make changes in a separate PR with added tests. We have updated this lifecycle method in other conversions, but they've been done with added tests and this is still the approach I'd prefer if this change is going to be made. If this is blocking, then please submit a full review requesting changes. This won't be merging until Monday, so let me know by then.

Comment (eatyourgreens) [reply app/talk/tags.jsx:45]: Not a blocking change at all but this will ensure that `getTags` only runs once on a change of props.

**Pull request merged**